### PR TITLE
adds a patch dataset

### PR DESCRIPTION
**Pull request by wyli**
part of  #606 #719 #1048 
this PR adds a simple patch-based dataset. it's useful when the number of patches per volume is fixed across an image dataset
(this is beyond v0.3 @Nic-Ma)

### Status
**ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/data/dataloader.py [hunk 1]
```diff
@@ -64,6 +65,11 @@ def __init__(
         timeout: float = 0.0,
         multiprocessing_context: Optional[Callable] = None,
     ) -> None:
+        if num_workers == 0:
+            # when num_workers > 0, random states are determined by worker_init_fn
+            # this is to make the behavior consistent when num_workers == 0
+            _seed = torch.empty((), dtype=torch.int64).random_(generator=None).item()
+            set_rnd(dataset, _seed)
         super().__init__(  # type: ignore[call-overload]
```
Comment (wyli) [inline monai/data/dataloader.py:73]: previously when `num_workers == 0`, the `worker_init_fn` is not used,  the loader's behaviour becomes different from those `num_workers > 0`. the modification here is to make them consistent. @Nic-Ma

File: monai/data/utils.py [hunk 2]
```diff
@@ -242,8 +242,24 @@ def worker_init_fn(worker_id: int) -> None:
 
     """
     worker_info = torch.utils.data.get_worker_info()
-    if hasattr(worker_info.dataset, "transform") and hasattr(worker_info.dataset.transform, "set_random_state"):
-        worker_info.dataset.transform.set_random_state(worker_info.seed % (2 ** 32))
+    set_rnd(worker_info.dataset, seed=worker_info.seed)
```
Comment (wyli) [inline monai/data/utils.py:245]: this is an enhancement to `set_random_state` for all the dataset properties, so that it supports nested dataset, such as: `Dataset(Dataset([data_1, data_2, ...], transform=random_transform_1), transform=random_transform_2)` @Nic-Ma it's useful for the patch dataset

File: monai/data/utils.py [hunk 3]
```diff
@@ -242,8 +242,24 @@ def worker_init_fn(worker_id: int) -> None:
...
+    """
+    Set seed or random state for all randomisable properties of obj.
+
+    Args:
+        seed: set the random state with an integer seed.
+    """
+    if not hasattr(obj, "__dict__"):
+        return seed  # no attribute
+    if hasattr(obj, "set_random_state"):
+        obj.set_random_state(seed=seed % (np.iinfo(np.int32).max + 1))
```
Comment (ericspod) [inline monai/data/utils.py:258]: If `seed` equals 2^31 then `seed % (np.iinfo(np.int32).max + 1)` would permit this value to be used which is one more than the maximum value for a 32-bit int (2^31-1). Did you perhaps mean `(seed + 1) % np.iinfo(np.int32).max`? If so before this line should be `seed = (seed + 1) % np.iinfo(np.int32).max` so that the correct in-bounds value is returned as well.

Comment (wyli) [reply monai/data/utils.py:258]: yeah, I just realised some issues here as well. let me look into this again.
for now: 
- `np.iinfo(np.int32).max == (2**31 - 1)`
- `np.random.RandomState` requires seed between `0` and `2**32 - 1` (inclusive)
I hope it's the same across the different OS. there're a few places in the codebase need to be improved...

**Review by ericspod — APPROVED** [review]
